### PR TITLE
Fixes forward AD codegen for multiple formulas

### LINKED ISSUE #67367: Multi-output forward grad codegen is wrong if output values are used
The current codegen allows the user to re-use the output's value in their formula. But when they are multiple outputs, this might already have been changed by the previous formula.

For example eigh is:
```yaml
- name: linalg_eigh(Tensor self, str UPLO="L") -> (Tensor eigenvalues, Tensor eigenvectors)
  self: eigh_backward(grads, self, /*eigenvectors=*/true, eigenvalues, eigenvectors)
  eigenvalues: eigh_jvp_eigenvalues(self_t, eigenvalues, eigenvectors)
  eigenvectors: eigh_jvp_eigenvectors(self_t, eigenvalues, eigenvectors)
```

And so the schematic codegen for it is (with some small edits):
```c++
  if (_any_has_forward_grad_eigenvalues) {
      // stuff
      auto eigenvalues_new_fw_grad = eigh_jvp_eigenvalues(self_t, eigenvalues, eigenvectors);
      eigenvalues._set_fw_grad(eigenvalues_new_fw_grad, /* level */ 0, /* is_inplace_op */ false);
  }
  if (_any_has_forward_grad_eigenvectors) {
      // stuff
      auto eigenvectors_new_fw_grad = eigh_jvp_eigenvectors(self_t, eigenvalues, eigenvectors);
      eigenvectors._set_fw_grad(eigenvectors_new_fw_grad, /* level */ 0, /* is_inplace_op */ false);
  }
```

You can see that by the time the eigenvectors formula uses eigenvalues, it already has its fw_grad populated. And so we're going to apply forward AD to `eigh_jvp_eigenvectorsv` and the output will be a weird Tensor whose forward grad also has a forward grad (which should not be possible!).

Note that having such nested forward grads actually lead to a deadlock on destruction of the level (depending on the order in which they are stored in the ordered set, so this is a non-consistent deadlock).

We should:
- Properly enforce that a forward grad cannot have a forward grad at the same level itself. You should not be able to do `make_dual(a, make_dual(b, c))` (note that this one does not usually deadlock due to the order in which they are registered on the level)
- We should fix the codegen to compute all the gradients first and then set all the gradients. This will ensure that we don't compute forward grad within the formula.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7

### PR DESCRIPTION
**Pull request by soulitzer**
Fixes #67367

- Adds check to make sure forward grad itself does not have forward grad at the same level
- Verify with `python test/test_ops.py -k test_forward_mode_AD_linalg_eigh_cpu_float64` that it fails the check before, but passes after the codegen update

Before:
```
  if (_any_has_forward_grad_eigenvalues) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_t = self_t_raw.defined() ? self_t_raw : at::zeros_like(toNonOptTensor(self));
      auto eigenvalues_new_fw_grad = eigh_jvp_eigenvalues(self_t, eigenvalues, eigenvectors);
      if (eigenvalues_new_fw_grad.defined()) {
        // The hardcoded 0 here will need to be updated once we support multiple levels.
        eigenvalues._set_fw_grad(eigenvalues_new_fw_grad, /* level */ 0, /* is_inplace_op */ false);
      }
  }
  if (_any_has_forward_grad_eigenvectors) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_t = self_t_raw.defined() ? self_t_raw : at::zeros_like(toNonOptTensor(self));
      auto eigenvectors_new_fw_grad = eigh_jvp_eigenvectors(self_t, eigenvalues, eigenvectors);
      if (eigenvectors_new_fw_grad.defined()) {
        // The hardcoded 0 here will need to be updated once we support multiple levels.
        eigenvectors._set_fw_grad(eigenvectors_new_fw_grad, /* level */ 0, /* is_inplace_op */ false);
      }
  }
```

After:
```
  c10::optional<at::Tensor> eigenvalues_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_eigenvalues) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_t = self_t_raw.defined() ? self_t_raw : at::zeros_like(toNonOptTensor(self));
      eigenvalues_new_fw_grad_opt = eigh_jvp_eigenvalues(self_t, eigenvalues, eigenvectors);
  }
  c10::optional<at::Tensor> eigenvectors_new_fw_grad_opt = c10::nullopt;
  if (_any_has_forward_grad_eigenvectors) {
      auto self_t_raw = toNonOptFwGrad(self);
      auto self_t = self_t_raw.defined() ? self_t_raw : at::zeros_like(toNonOptTensor(self));
      eigenvectors_new_fw_grad_opt = eigh_jvp_eigenvectors(self_t, eigenvalues, eigenvectors);
  }
  if (eigenvalues_new_fw_grad_opt.has_value() && eigenvalues_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    eigenvalues._set_fw_grad(eigenvalues_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ false);
  }
  
  if (eigenvectors_new_fw_grad_opt.has_value() && eigenvectors_new_fw_grad_opt.value().defined()) {
    // The hardcoded 0 here will need to be updated once we support multiple levels.
    eigenvectors._set_fw_grad(eigenvectors_new_fw_grad_opt.value(), /* level */ 0, /* is_inplace_op */ false);
  }
```

### CODE REVIEW

File: torch/csrc/autograd/autograd_meta.cpp [hunk 1]
```diff
@@ -102,6 +102,8 @@ namespace {
 // This function is will ensure that the fw_grad_ is properly a view of the base for inplace ops on
 // Tensors that do not have forward grad originally.
 void AutogradMeta::set_fw_grad(const at::TensorBase& new_grad_base, const at::TensorBase& self_base, uint64_t level, bool is_inplace_op) {
+  TORCH_CHECK(!new_grad_base._fw_grad(level).defined(), "Setting a forward grad that "
```
Comment (soulitzer) [inline torch/csrc/autograd/autograd_meta.cpp:105]: Maybe this should be internal assert?

File: aten/src/ATen/core/TensorBase.h [hunk 2]
```diff
@@ -636,6 +636,14 @@ class TORCH_API TensorBase {
     return impl_->requires_grad();
   }
 
+  // The Forward AD API functions below are low level and are not to be used by end
+  // users who should use the API provided in torch/csrc/autograd.h
+
+  /// This function returns the forward gradient for this Tensor at the given level.
+  const Tensor& _fw_grad(uint64_t level) const {
```
Comment (albanD) [inline aten/src/ATen/core/TensorBase.h:643]: Why was this not here before?
Are there other fw grad methods? If so, this one should be next to them.

File: torch/csrc/autograd/autograd_meta.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/autograd_meta.cpp:105]: No, the user could try to do `make_dual(foo, make_dual(foo, foo))`. We should have a test for that btw.

**Review by albanD — COMMENTED** [review]
Sounds good.
Missing a test and good to go!

File: aten/src/ATen/core/TensorBase.h (hunk 2 above)
Comment (soulitzer) [reply aten/src/ATen/core/TensorBase.h:643]: Not too sure why it wasn't here before. Probably because it wasn't needed anywhere. I've added the other one too now for completeness.

Comment (albanD) [reply aten/src/ATen/core/TensorBase.h:643]: We are already using them successfully in https://github.com/pytorch/pytorch/blob/533e72e0a4a7595d157c779d1f001ce78b29f80e/tools/autograd/gen_variable_type.py#L331 and https://github.com/pytorch/pytorch/blob/533e72e0a4a7595d157c779d1f001ce78b29f80e/torch/csrc/autograd/FunctionsManual.cpp#L54 so I don't think we should need them here.

Comment (soulitzer) [reply aten/src/ATen/core/TensorBase.h:643]: In those cases those are Tensors and not TensorBase. Adding this will address the TensorBase case.

Comment (albanD) [reply aten/src/ATen/core/TensorBase.h:643]: Ho I forgot about TensorBase, good point!

**Review by albanD — APPROVED** [review]
SGTM
